File: app/src/main/java/com/nominacopro/domain/calculator/HourCalculator.kt
package com.nominacopro.domain.calculator

import com.nominacopro.domain.law.ColombiaLaborLaw2026
import com.nominacopro.domain.model.DayType
import com.nominacopro.domain.model.HourBreakdown
import com.nominacopro.domain.model.WorkDayEntry
import java.time.LocalDate
import java.time.LocalTime

/**
 * Calcula desglose de horas según CST + Ley 2466/2025.
 * Descuenta 1 h de almuerzo (12:00–13:00) si el turno la cubre.
 */
object HourCalculator {

    private const val LUNCH_START = 12 * 60
    private const val LUNCH_END = 13 * 60
    private const val NIGHT_START = 19 * 60
    private const val NIGHT_END_MIDNIGHT = 24 * 60
    private const val NIGHT_END_MORNING = 6 * 60

    fun calculate(
        entry: WorkDayEntry,
        dailyHours: Int,
        isRestDay: Boolean,
    ): HourBreakdown {
        if (entry.dayType == DayType.FESTIVO_DOMINICAL || entry.dayType == DayType.FESTIVO_NOCTURNO || isRestDay) {
            return calculateRestDay(entry, dailyHours)
        }
        return calculateWeekday(entry, dailyHours)
    }

    private fun calculateWeekday(entry: WorkDayEntry, dailyHours: Int): HourBreakdown {
        val startMin = toMinutes(entry.start)
        var endMin = toMinutes(entry.end)
        if (endMin <= startMin) endMin += 24 * 60

        val lunch = lunchOverlap(startMin, endMin)
        val totalHours = (endMin - startMin - lunch) / 60.0
        val jornada = dailyHours.toDouble()

        val nightHours = nightMinutes(startMin, endMin - lunch) / 60.0
        val dayHours = (totalHours - nightHours).coerceAtLeast(0.0)

        var extraDiurna = (totalHours - jornada).coerceAtLeast(0.0)
        var extraNocturna = 0.0
        var nocturnaOrd = nightHours.coerceAtMost(jornada)
        var normalDiurna = (jornada - nocturnaOrd).coerceAtLeast(0.0)

        if (totalHours > jornada) {
            val over = totalHours - jornada
            extraNocturna = (nightHours - nocturnaOrd).coerceAtMost(over)
            extraDiurna = over - extraNocturna
            normalDiurna = (dayHours - extraDiurna).coerceAtLeast(0.0)
            nocturnaOrd = nightHours - extraNocturna
        } else {
            normalDiurna = dayHours.coerceAtMost(jornada - nocturnaOrd)
        }

        return HourBreakdown(
            normalDiurna = round2(normalDiurna),
            nocturnaOrdinaria = round2(nocturnaOrd),
            extraDiurna = round2(extraDiurna),
            extraNocturna = round2(extraNocturna),
        )
    }

    private fun calculateRestDay(entry: WorkDayEntry, dailyHours: Int): HourBreakdown {
        val startMin = toMinutes(entry.start)
        var endMin = toMinutes(entry.end)
        if (endMin <= startMin) endMin += 24 * 60
        val lunch = lunchOverlap(startMin, endMin)
        val totalHours = (endMin - startMin - lunch) / 60.0
        val nightHours = nightMinutes(startMin, endMin - lunch) / 60.0
        val dayHours = (totalHours - nightHours).coerceAtLeast(0.0)
        val jornada = dailyHours.toDouble()

        val dominicalDiurna = dayHours.coerceAtMost(jornada)
        val dominicalNocturna = nightHours.coerceAtMost(jornada - dominicalDiurna)
        val extraDomDiu = (dayHours - dominicalDiurna).coerceAtLeast(0.0)
        val extraDomNoc = (nightHours - dominicalNocturna).coerceAtLeast(0.0)

        return HourBreakdown(
            dominicalDiurna = round2(dominicalDiurna),
            dominicalNocturna = round2(dominicalNocturna),
            extraDominicalDiurna = round2(extraDomDiu),
            extraDominicalNocturna = round2(extraDomNoc),
        )
    }

    private fun lunchOverlap(start: Int, end: Int): Int {
        val overlapStart = maxOf(start, LUNCH_START)
        val overlapEnd = minOf(end, LUNCH_END)
        return (overlapEnd - overlapStart).coerceAtLeast(0)
    }

    private fun nightMinutes(start: Int, end: Int): Int {
        var total = 0
        val ranges = listOf(
            0 to NIGHT_END_MORNING,
            NIGHT_START to NIGHT_END_MIDNIGHT,
            NIGHT_END_MIDNIGHT to NIGHT_END_MIDNIGHT + NIGHT_END_MORNING,
        )
        for ((a, b) in ranges) {
            val ol = maxOf(start, a)
            val oh = minOf(end, b)
            if (oh > ol) total += oh - ol
        }
        return total
    }

    private fun toMinutes(time: LocalTime): Int = time.hour * 60 + time.minute

    private fun round2(v: Double): Double = kotlin.math.round(v * 100) / 100.0
}

File: web/lib/payroll/hourCalculator.ts
import type { HourBreakdown, WorkDayEntry } from "./models";
import { parseTimeToMinutes } from "./dates";

const LUNCH_START = 12 * 60;
const LUNCH_END = 13 * 60;
const NIGHT_START = 19 * 60;
const NIGHT_END_MORNING = 6 * 60;
const MINUTES_PER_DAY = 24 * 60;

function round2(v: number): number {
  return Math.round(v * 100) / 100;
}

function isNightAt(minute: number): boolean {
  const m = minute % MINUTES_PER_DAY;
  return m < NIGHT_END_MORNING || m >= NIGHT_START;
}

function workedMinuteRanges(startMin: number, endMin: number): [number, number][] {
  if (endMin <= startMin) return [];
  if (endMin <= LUNCH_START || startMin >= LUNCH_END) {
    return [[startMin, endMin]];
  }
  const ranges: [number, number][] = [];
  if (startMin < LUNCH_START) {
    ranges.push([startMin, Math.min(LUNCH_START, endMin)]);
  }
  if (endMin > LUNCH_END) {
    ranges.push([Math.max(LUNCH_END, startMin), endMin]);
  }
  return ranges;
}

function chronologySegments(startMin: number, endMin: number): [number, boolean][] {
  const result: [number, boolean][] = [];
  for (const [segStart, segEnd] of workedMinuteRanges(startMin, endMin)) {
    const boundaries = new Set<number>([segStart, segEnd]);
    let dayAnchor = Math.floor(segStart / MINUTES_PER_DAY) * MINUTES_PER_DAY - MINUTES_PER_DAY;
    while (dayAnchor <= segEnd) {
      boundaries.add(dayAnchor + NIGHT_END_MORNING);
      boundaries.add(dayAnchor + NIGHT_START);
      boundaries.add(dayAnchor + MINUTES_PER_DAY);
      dayAnchor += MINUTES_PER_DAY;
    }
    const sorted = Array.from(boundaries)
      .filter((b) => b >= segStart && b <= segEnd)
      .sort((a, b) => a - b);
    for (let i = 0; i < sorted.length - 1; i++) {
      const from = sorted[i];
      const to = sorted[i + 1];
      if (to > from) {
        result.push([(to - from) / 60, isNightAt(from)]);
      }
    }
  }
  return result;
}

function allocateHours(
  segments: [number, boolean][],
  jornada: number,
  restDay: boolean,
): HourBreakdown {
  let remainingOrdinary = jornada;
  let normalDiurna = 0;
  let nocturnaOrd = 0;
  let extraDiurna = 0;
  let extraNocturna = 0;
  let dominicalDiurna = 0;
  let dominicalNocturna = 0;
  let extraDominicalDiurna = 0;
  let extraDominicalNocturna = 0;

  for (const [hours, isNight] of segments) {
    let left = hours;
    const ordinary = Math.min(left, Math.max(remainingOrdinary, 0));
    if (ordinary > 0) {
      if (restDay) {
        if (isNight) dominicalNocturna += ordinary;
        else dominicalDiurna += ordinary;
      } else {
        if (isNight) nocturnaOrd += ordinary;
        else normalDiurna += ordinary;
      }
      remainingOrdinary -= ordinary;
      left -= ordinary;
    }
    if (left > 0) {
      if (restDay) {
        if (isNight) extraDominicalNocturna += left;
        else extraDominicalDiurna += left;
      } else {
        if (isNight) extraNocturna += left;
        else extraDiurna += left;
      }
    }
  }

  return {
    normalDiurna: round2(normalDiurna),
    nocturnaOrdinaria: round2(nocturnaOrd),
    extraDiurna: round2(extraDiurna),
    extraNocturna: round2(extraNocturna),
    dominicalDiurna: round2(dominicalDiurna),
    dominicalNocturna: round2(dominicalNocturna),
    extraDominicalDiurna: round2(extraDominicalDiurna),
    extraDominicalNocturna: round2(extraDominicalNocturna),
  };
}

export function calculateHours(
  entry: WorkDayEntry,
  dailyHours: number,
  isRestDay: boolean,
): HourBreakdown {
  const startMin = parseTimeToMinutes(entry.start);
  let endMin = parseTimeToMinutes(entry.end);
  if (endMin <= startMin) endMin += MINUTES_PER_DAY;

  const restDay =
    entry.dayType === "FESTIVO_DOMINICAL" ||
    entry.dayType === "FESTIVO_NOCTURNO" ||
    isRestDay;
  const segments = chronologySegments(startMin, endMin);
  return allocateHours(segments, dailyHours, restDay);
}
